src/clip.sol

```diff
@@ -0,0 +1,472 @@
+            vat.flux(ilk, address(this), who, slice);
+
+            // Do external call (if data is defined) but to be
+            // extremely careful we don't allow to do it to the two
+            // contracts which the Clipper needs to be authorized
+            DogLike dog_ = dog;
+            if (data.length > 0 && who != address(vat) && who != address(dog_)) {
+                ClipperCallee(who).clipperCall(msg.sender, owe, slice, data);
+            }
+
+            // Get DAI from caller
+            vat.move(msg.sender, vow, owe);
+
+            // Removes Dai out for liquidation from accumulator
+            dog_.digs(ilk, lot == 0 ? tab + owe : owe);
+        }
+
+        if (lot == 0) {
+            _remove(id);
+        } else if (tab == 0) {
+            vat.flux(ilk, address(this), usr, lot);
+            _remove(id);
+        } else {
+            sales[id].tab = tab;
+            sales[id].lot = lot;
+        }
+
+        emit Take(id, max, price, owe, tab, lot, usr);
+    }
+
+    function _remove(uint256 id) internal {
+        uint256 _move    = active[active.length - 1];
+        if (id != _move) {
+            uint256 _index   = sales[id].pos;
+            active[_index]   = _move;
+            sales[_move].pos = _index;
+        }
+        active.pop();
+        delete sales[id];
+    }
+
+    // The number of active auctions
+    function count() external view returns (uint256) {
+        return active.length;
+    }
+
+    // Return the entire array of active auctions
+    function list() external view returns (uint256[] memory) {
+        return active;
+    }
+
+    // Externally returns boolean for if an auction needs a redo and also the current price
+    function getStatus(uint256 id) external view returns (bool needsRedo, uint256 price, uint256 lot, uint256 tab) {
+        // Read auction data
+        address usr = sales[id].usr;
+        uint96  tic = sales[id].tic;
+
+        bool done;
+        (done, price) = status(tic, sales[id].top);
+
+        needsRedo = usr != address(0) && done;
+        lot = sales[id].lot;
+        tab = sales[id].tab;
+    }
+
+    // Internally returns boolean for if an auction needs a redo
+    function status(uint96 tic, uint256 top) internal view returns (bool done, uint256 price) {
+        price = calc.price(top, sub(block.timestamp, tic));
+        done  = (sub(block.timestamp, tic) > tail || rdiv(price, top) < cusp);
+    }
+
+    // Public function to update the cached dust*chop value.
+    function upchost() external {
+        (,,,, uint256 _dust) = VatLike(vat).ilks(ilk);
+        chost = wmul(_dust, dog.chop(ilk));
+    }
+
+    // Cancel an auction during ES or via governance action.
+    function yank(uint256 id) external auth lock {
+        require(sales[id].usr != address(0), "Clipper/not-running-auction");
+        dog.digs(ilk, sales[id].tab);
+        vat.flux(ilk, address(this), msg.sender, sales[id].lot);
+        _remove(id);
+        emit Yank(id);
+    }
```

Comment:
I think we should start including this on every contract to reduce the need for this `upchost`-style caching with big structs:

```
function _peek(bytes32 addr) external pure returns (bytes32 b) {
    assembly {
        b := mload(addr)
    }
}
```

Reply:
As I said in a prev comment, we should discuss this for liquidations 2.1. Maybe it is just easier to add all the methods manually when we have a struct, but this is also a cool idea that as you mentioned works in the meanwhile we do not want to protect an onchain read for a value.